jvb/src/main/kotlin/org/jitsi/videobridge/relay/RelayMessageTransport.kt

```diff
@@ -0,0 +1,426 @@
+import org.jitsi.videobridge.message.EndpointConnectionStatusMessage
+import org.jitsi.videobridge.message.EndpointMessage
+import org.jitsi.videobridge.message.EndpointStats
+import org.jitsi.videobridge.message.RemoveReceiverMessage
+import org.jitsi.videobridge.message.SelectedEndpointMessage
+import org.jitsi.videobridge.message.SelectedEndpointsMessage
+import org.jitsi.videobridge.message.ServerHelloMessage
+import org.jitsi.videobridge.message.SourceVideoTypeMessage
+import org.jitsi.videobridge.message.VideoTypeMessage
+import org.jitsi.videobridge.websocket.ColibriWebSocket
+import org.json.simple.JSONObject
+import java.net.URI
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.atomic.AtomicLong
+import java.util.function.Supplier
+import java.util.stream.Collectors
+
+/**
+ * Handles the functionality related to sending and receiving COLIBRI messages
+ * for a [Relay].
+ */
+class RelayMessageTransport(
+    private val relay: Relay,
+    private val statisticsSupplier: Supplier<Videobridge.Statistics>,
+    private val eventHandler: EndpointMessageTransportEventHandler,
+    parentLogger: Logger
+) : AbstractEndpointMessageTransport(parentLogger), ColibriWebSocket.EventHandler {
+    /**
+     * The last connected/accepted web-socket by this instance, if any.
+     */
+    private var webSocket: ColibriWebSocket? = null
+
+    /**
+     * For active websockets, the URL that was connected to.
+     */
+    private var url: String? = null
+
+    /**
+     * An active websocket client.
+     */
+    private var outgoingWebsocket: WebSocketClient? = null
+
+    /**
+     * Use to synchronize access to [.webSocket]
+     */
+    private val webSocketSyncRoot = Any()
+    private val numOutgoingMessagesDropped = AtomicInteger(0)
+
+    /**
+     * The number of sent message by type.
+     */
+    private val sentMessagesCounts: MutableMap<String, AtomicLong> = ConcurrentHashMap()
+
+    init { logger.addContext("relay-id", relay.id) }
+
+    /**
+     * Connect the bridge channel message to the websocket URL specified
+     */
+    fun connectTo(url: String) {
+        if (this.url != null && this.url == url) {
+            return
+        }
+        this.url = url
+
+        doConnect()
+    }
+
+    private fun doConnect() {
+        val url = this.url ?: throw IllegalStateException("Cannot connect Relay transport when no URL set")
+
+        webSocket = ColibriWebSocket(relay.id, this)
+        outgoingWebsocket = WebSocketClient().also {
+            it.start()
+            it.connect(webSocket, URI(url), ClientUpgradeRequest())
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    override fun notifyTransportChannelConnected() {
+        relay.relayMessageTransportConnected()
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    override fun clientHello(message: ClientHelloMessage): BridgeChannelMessage? {
```

Comment:
I don't think we need the ClientHello/ServerHello exchange. The websocket events should be sufficient

Reply:
The one thing it provides is the peer JVB version, which would let us know if the servers are out of sync for some reason.

Reply:
Should we log a warning if that's the case?

Reply:
Not a bad plan.